lib/trace_mgr.js

```diff
@@ -0,0 +1,611 @@
+    if (!entry) {
+        return buildTraceResult('no-sample');
+    }
+
+    //build up trace name 
+    var traceBucketName = `${emitKind}_${stk.fbin}_bucket-${entry.bucketId}`;
+    var traceDirName = `trace-${entry.traceCtr}`;
+    var traceName = traceBucketName + path.sep + traceDirName + path.sep;
+
+    //invoke the trace writer and remote manager if needed
+    var resolvedPath = createTraceLogTarget(traceName);
+    if (!resolvedPath) {
+        return buildTraceResult('fail');
+    }
+
+    var action = function () {
+        try {
+            process.stderr.write(`    Write error trace to: ${resolvedTracePath}\n`);
+            global.emitTTDLog(resolvedTracePath);
+
+            if(emitOptions.remoteTraceManagerObj && emitOptions.remoteTraceManagerObj.uploadTraceAsync) {
+                emitOptions.remoteTraceManagerObj.uploadTraceAsync(resolvedPath);
+            }
+        } catch(ex) {
+            process.stderr.write('Failed in emitTrace with: ' + ex);
+        }
+    }
+
+    //update the bin stats as needed
+    updateEmitInfo(entry, resolvedPath);
+
+    return buildTraceResult('success', action);
+}
+
+//create a directory for emitting the trace (if possible) and return it
+function createTraceLogTarget(tracename) {
+    var traceRootDir = emitOptions.localTraceDirectory;
+    if (traceRootDir === undefined) {
+        traceRootDir = path.dirname(process.mainModule.filename);
+    }
+    resolvedTracePath = path.resolve(traceRootDir, '_diagnosticTraces', tracename);
+
+    //ensure directory exists and is empty...
+    var ok = ensureTraceTarget(resolvedTracePath);
+    if (!ok) {
+        return undefined;
+    }
+
+    return resolvedTracePath;
+}
+
+//ensure directory exists and is empty...
+function ensureTraceTarget(pth) {
+    //I don't like this and don't want it to be happening here so I am going to bail
+    if (!path.isAbsolute(pth)) {
+        return false;
+    }
+
+    var okdir = createTargetDirectory(pth);
+    if (!okdir) {
+        return false;
+    }
+
+    return deleteTargetDirectoryContents(pth);
+}
+
+function createTargetDirectory(pth) {
+    //see if it just exists and, if so, just return true
+    try {
+        fs.accessSync(pth, fs.constants.R_OK | fs.constants.W_OK | fs.constants.X_OK);
+        return true;
+    }
+    catch (ei) { ; }
+
+    //walk up the directory to see where the first valid part of the path is
+    var prefixPath = pth;
+    var suffixPaths = [];
+    var baseFound = false;
+    do {
+        //check for bad prefix 
+        if (prefixPath === path.dirname(prefixPath)) {
+            process.stderr.write(`Failed creating trace directory (bad prefix): ${pth} -> ${prefixPath}\n`);
+            return false;
+        }
+
+        suffixPaths.push(path.basename(prefixPath)); //reverse order 
+        prefixPath = path.dirname(prefixPath);
+
+        try {
+            fs.accessSync(prefixPath, fs.constants.R_OK | fs.constants.W_OK | fs.constants.X_OK);
```

Comment:
This would pass even if prefixPath was a file, no? Maybe you should do fs.stat here instead and confirm it's a directory?